src/ref.js

```diff
@@ -0,0 +1,34 @@
+import isPlainObject from './helpers/isPlainObject';
+import wrapValidator from './helpers/wrapValidator';
+
+function isNewRef(prop) {
+  if (!isPlainObject(prop)) {
+    return false;
+  }
+  const ownProperties = Object.getOwnPropertyNames(prop);
+  return ownProperties.length === 1 && ownProperties[0] === 'current';
+}
+
+function ref(props, propName, componentName) {
+  const { [propName]: value } = props;
+
+  if (value == null || typeof value === 'function' || isNewRef(value)) {
```

Comment:
should this ensure that when value is a function, it doesn't extend `React.Component` or `React.PureComponent`?

Reply:
This comment got buried, but I'm not sure we need to go that far. I'm thinking that for these prop types, it's not worth the time to figure out if the function extends either of the React.Components

Reply:
Up to you, it doesn’t have to be a blocker - but in general I’ve found that every level of validation is always worth the time, especially in something non-production like propTypes :-)

Reply:
That's fair. Looks like it could be something like:

```js
React.Component.isPrototypeOf(value);
```

If we wanted to get real crazy we could check the arity of the function as well (since it should have a single param).

Reply:
i'd be happy with those as improvements in a followup PR :-)

Reply:
Will do. I also made a slight mistake on the documentation (I think `ref` needs to be invoked when using it), and will update

Reply:
No worries, that I fixed prior to merging :-)